### PR TITLE
DebugMode using Particles & Comments

### LINKED ISSUE #309: Add a debug mode
It should, for example, spawn particles where units are server-side so that we see if they match their positions in the client. Other debug info should be added too

### PR DESCRIPTION
**Pull request by lzardy**
Adds the !debugmode command which has the option to draw the player's server-side collision radius and waypoints. This can be expanded to other units such as projectiles and minions.
Streamlined the process of adding, removing, and networking particles.
Added more variables to particles related to networking.
Cleaned the GameObject.Move function and added some functions related to movement.
Cleaned the NotifyFXCreateGroup function and used new Particle variables.
Added comments to GameObject and Particle.
Shifted location of adding players in Game so that PacketNotifier can never cause an exception.

Resolves #309

### CODE REVIEW

**Comment by danil179** [comment]
I'm merging PR #971 as #966 seems to not occur anymore (they probably have fixed it). That will probably create some merge conflicts. So after I'll end merging you will need to simply rebase (it shouldn't take long to rebase it and it shouldn't be complicated, be sure to keep always another backup branch in case something goes wrong).

**Comment by lzardy** [comment]
Rebase successful.

**Comment by lzardy** [comment]
Closing to test travis re-build on re-PR.

File: GameServerLib/Chatbox/Commands/DebugModeCommand.cs [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+            var champions = Game.ObjectManager.GetAllChampions();
+
+            _logger.Debug($"Started debugging " + champions.Count + " champions. Your Debug Circle Radius: " + "(1 / 100) * " + _userChampion.CollisionRadius + " = " + "(" + (1f / 100f) + ") * " + _userChampion.CollisionRadius + " = " + circlesize);
+            var startdebugmsg = $"Started debugging " + champions.Count + " champions. Your Debug Circle Radius: " + "(1 / 100) * " + _userChampion.CollisionRadius + " = " + "(" + (1f / 100f) + ") * " + _userChampion.CollisionRadius + " = " + circlesize;
+            ChatCommandManager.SendDebugMsgFormatted(DebugMsgType.NORMAL, startdebugmsg);
+
+            // Creates a blue flashing highlight around your unit
+            _game.PacketNotifier.NotifyCreateUnitHighlight(userId, _userChampion);
+
+            for (var i = 0.0f; i < 86400.00f && _debugMode == 2; i += 0.1f) //debug for 24 hours, drawing every 100 milliseconds
```
Comment (danil179) [inline GameServerLib/Chatbox/Commands/DebugModeCommand.cs:224]: This is not how it should be done.

File: GameServerLib/GameObjects/GameObject.cs [hunk 2]
```diff
@@ -91,33 +94,43 @@ public virtual void OnCollision(IGameObject collider)
...
 
-        public void Move(float diff)
+        /// <summary>
+        /// Moves the object to its specified waypoints, updating its coordinate.
+        /// </summary>
+        /// <param name="diff">The amount of milliseconds the object is supposed to move</param>
+        public void Move(float diff, bool useTarget = false)
         {
+            // no waypoints remained - clear the Waypoints
+            if (WaypointIndex >= Waypoints.Count)
```
Comment (danil179) [inline GameServerLib/GameObjects/GameObject.cs:104]: Use your function `IsPathEnded`.

File: GameServerLib/GameObjects/GameObject.cs [hunk 3]
```diff
@@ -91,33 +94,43 @@ public virtual void OnCollision(IGameObject collider)
...
+            {
+                Waypoints.RemoveAll(v => v != Waypoints[0]);
```
Comment (danil179) [inline GameServerLib/GameObjects/GameObject.cs:106]: Calling RemoveAll every time is probably not doing good to the CPU. Maybe add to the condition ` && Waypoints.Count > 1`.

**Review by danil179 — APPROVED** [review]
Approved. Some minor comments.

**Pull request merged**